runtime/compiler/ilgen/IlGenerator.cpp

```diff
@@ -1004,15 +1004,17 @@ TR_J9ByteCodeIlGenerator::genExceptionHandlers(TR::Block * lastBlock)
          }
       TR::Block *catchBlock   = handlerInfo._catchBlock;
       TR::Block *restartBlock = _blocksToInline? _blocksToInline->getGeneratedRestartTree()->getEnclosingBlock() : NULL;
-      uint8_t precedingOpcode = _code[handlerInfo._startIndex-1];
-      TR_J9ByteCode precedingBytecode = convertOpCodeToByteCodeEnum(precedingOpcode);
-      uint8_t lastOpcode = _code[handlerInfo._endIndex];
-      TR_J9ByteCode lastBytecode = convertOpCodeToByteCodeEnum(lastOpcode);
-      bool isSynchronizedRegion = (precedingBytecode == J9BCmonitorenter && lastBytecode == J9BCmonitorexit);
-      if (isSynchronizedRegion) // monitorenter preceding try region that ends in monitorexit means synchronized
-         catchBlock->setIsSynchronizedHandler();
-
-      for (uint16_t j = handlerInfo._startIndex; j <= handlerInfo._endIndex; ++j)
+      if (handlerInfo._startIndex)
```

Comment:
What is the significance of this extra test that is added ?

Reply:
> What is the significance of this extra test that is added ?

This code is looking for try regions that starts right after a mon-enter and ends on a mon-exit so it can set a flag on the catch handler bock. When a try region starts at bytecode index 0 it's not possible that the try region is preceded by a mon-enter. Allowing this code to look for a mon-enter when the `_startIndex` is 0 would allow a negative index into the array and relies on that memory being addressable (which is not guaranteed) to prevent a SEGV. It also leaves the possibility that the memory at array index -1 would just happen to have the value matching J9BCmonitorenter so we might incorrectly set the flag. 

So this new test stops us from looking for the condition where we would want to set the flag under a scenario where the condition is really not possible, and by doing this new test we protect against a SEGV or incorrectly setting the flag based on random memory contents above the bytecode array.

At first I had changed the uint16_t to uint32_t and that resulted in very large positive offsets when subtracting 1 from a `_startIndex` of 0, this would cause a SEGV in code as described above. Now that I have changed the type to a signed 32bit int this crash is far less likely to occur, but the issue is still possible without this new test.

Reply:
I added a comment to explain the need for this new test.